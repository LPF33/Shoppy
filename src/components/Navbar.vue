<template>
  <nav v-if="isLoggedIn">
    <router-link :to="{ name: 'Main' }">Main</router-link>
    <router-link :to="{ name: 'Shoppy' }">Shoppy</router-link>
    <router-link :to="{ name: 'Cook' }">Cooky</router-link>
  </nav>
  <nav v-else>
    <router-link :to="{ name: 'Login' }">Login</router-link>
  </nav>
</template>

<script>
export default {
  name: "Navbar",
  computed: {
    isLoggedIn: function () {
      return this.$store.state.isAuthenticated;
    },
  },
};
</script>

<style>
nav {
  width: 100%;
  padding: 10px;
  display: flex;
  justify-content: center;
  background: #232323;
}

nav a {
  font-weight: bold;
  color: #2c3e50;
  border: 2px solid rgb(184, 184, 184);
  padding: 10px;
  width: 150px;
  text-align: center;
  margin: 0 10px;
  color: #ffeda3;
  text-decoration: none;
}

nav a.router-link-exact-active {
  color: #39a9cb;
}
</style>
