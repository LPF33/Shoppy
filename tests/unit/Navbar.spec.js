import { mount, flushPromises } from "@vue/test-utils";
import { createRouter, createWebHistory } from "vue-router";
import { routes } from "@/router/index.js";
import Navbar from "@/components/Navbar";

let $store = {};

beforeEach(() => {
  $store = {
    state: {
      error: false,
      isAuthenticated: false,
    },
  };
});

const router = createRouter({
  history: createWebHistory(),
  routes,
});

const mountNavbar = () => {
  return mount(Navbar, {
    global: {
      plugins: [router],
      mocks: { $store },
    },
  });
};

describe("Testing the Navbar component", () => {
  it("When user is not logged in, Navbar should have on router-link", async () => {
    router.push("/");
    await router.isReady();
    const wrapper = mountNavbar();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find("nav").exists()).toBeTruthy();
    expect(wrapper.findAll("a")).toHaveLength(1);
    expect(wrapper.findAll("a")[0].attributes("href")).toBe("/login");
    expect(wrapper.findAll("a")[0].text()).toBe("Login");
  });

  it("When user is logged in, Navbar should render three router-links", async () => {
    $store.state.isAuthenticated = true;
    router.push("/");
    await router.isReady();
    const wrapper = mountNavbar();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find("nav").exists()).toBeTruthy();
    expect(wrapper.findAll("a")).toHaveLength(3);
    expect(wrapper.findAll("a")[0].attributes("href")).toBe("/");
    expect(wrapper.findAll("a")[0].text()).toBe("Main");
    expect(wrapper.findAll("a")[1].attributes("href")).toBe("/shoppy");
    expect(wrapper.findAll("a")[1].text()).toBe("Shoppy");
    expect(wrapper.findAll("a")[2].attributes("href")).toBe("/cooky");
    expect(wrapper.findAll("a")[2].text()).toBe("Cooky");
    expect(wrapper.findAll(".router-link-active")).toHaveLength(1);
    expect(wrapper.find(".router-link-active").attributes("href")).toBe("/");
  });

  it("User is logged in, navigate to Shoppy-Route, Navbar should render accordingly", async () => {
    $store.state.isAuthenticated = true;
    router.push("/shoppy");
    await router.isReady();
    const wrapper = mountNavbar();

    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find("nav").exists()).toBeTruthy();
    expect(wrapper.findAll("a")).toHaveLength(3);
    expect(wrapper.findAll(".router-link-active")).toHaveLength(1);
    expect(wrapper.find(".router-link-active").attributes("href")).toBe(
      "/shoppy"
    );
  });

  it("User is logged in, navigate to Cooky-Route, Navbar should render accordingly", async () => {
    $store.state.isAuthenticated = true;
    router.push("/cooky");
    await router.isReady();
    const wrapper = mountNavbar();

    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find("nav").exists()).toBeTruthy();
    expect(wrapper.findAll("a")).toHaveLength(3);
    expect(wrapper.findAll(".router-link-active")).toHaveLength(1);
    expect(wrapper.find(".router-link-active").attributes("href")).toBe(
      "/cooky"
    );
  });

  it("When error occurs, div with CSS class error-nav gets rendered", async () => {
    $store.state.error = true;
    await router.isReady();
    const wrapper = mountNavbar();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find(".error-nav").exists()).toBeTruthy();
    expect(wrapper.find(".error-nav").text()).toBe("❗");
  });

  it("Click on router-link with name Shoppy navigates to '/shoppy'", async () => {
    $store.state.isAuthenticated = true;
    await router.isReady();
    const wrapper = mountNavbar();

    const shoppyLink = wrapper.findAll("a")[1];
    expect(shoppyLink.attributes("href")).toBe("/shoppy");
    shoppyLink.trigger("click");
    await flushPromises();
    expect(wrapper.vm.$route.path).toBe("/shoppy");
  });

  it("Click on router-link with name Cooky navigates to '/cooky'", async () => {
    $store.state.isAuthenticated = true;
    await router.isReady();
    const wrapper = mountNavbar();

    const cookyLink = wrapper.findAll("a")[2];
    expect(cookyLink.attributes("href")).toBe("/cooky");
    cookyLink.trigger("click");
    await flushPromises();
    expect(wrapper.vm.$route.path).toBe("/cooky");
  });
});
