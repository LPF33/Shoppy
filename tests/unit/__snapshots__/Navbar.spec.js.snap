// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Navbar component User is logged in, navigate to Cooky-Route, Navbar should render accordingly 1`] = `
<nav><a href="/" class="">Main</a><a href="/shoppy" class="">Shoppy</a><a href="/cooky" class="router-link-active router-link-exact-active" aria-current="page">Cooky</a></nav>
<!--v-if-->
`;

exports[`Testing the Navbar component User is logged in, navigate to Shoppy-Route, Navbar should render accordingly 1`] = `
<nav><a href="/" class="">Main</a><a href="/shoppy" class="router-link-active router-link-exact-active" aria-current="page">Shoppy</a><a href="/cooky" class="">Cooky</a></nav>
<!--v-if-->
`;

exports[`Testing the Navbar component When error occurs, div with CSS class error-nav gets rendered 1`] = `
<nav><a href="/login" class="">Login</a></nav>
<div class="error-nav">‚ùó</div>
`;

exports[`Testing the Navbar component When user is logged in, Navbar should render three router-links 1`] = `
<nav><a aria-current="page" href="/" class="router-link-active router-link-exact-active">Main</a><a href="/shoppy" class="">Shoppy</a><a href="/cooky" class="">Cooky</a></nav>
<!--v-if-->
`;

exports[`Testing the Navbar component When user is not logged in, Navbar should have on router-link 1`] = `
<nav><a href="/login" class="">Login</a></nav>
<!--v-if-->
`;
